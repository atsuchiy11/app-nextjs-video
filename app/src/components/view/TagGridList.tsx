import * as React from 'react'
import { makeStyles, Theme } from '@material-ui/core/styles'

import Box from '@material-ui/core/Box'
import Chip from '@material-ui/core/Chip'
import Divider from '@material-ui/core/Divider'
import IconButton from '@material-ui/core/IconButton'
import ArrowBackIosIcon from '@material-ui/icons/ArrowBackIos'
import ArrowForwardIosIcon from '@material-ui/icons/ArrowForwardIos'

import { animateScroll as scroll } from 'react-scroll'
import { useAppContext } from 'src/foundations/AppProvider'
import { useTags } from 'src/foundations/hooks'

const useStyles = makeStyles((theme: Theme) => ({
	wraper: {
		'& .MuiBox-root::-webkit-scrollbar': {
			display: 'none',
		},
		display: 'flex',
	},
	root: {
		display: 'flex',
		overflowX: 'auto',
	},
	tag: {
		margin: theme.spacing(0.5),
		backgroundColor: theme.palette.background.paper,
	},
	divider: {
		margin: theme.spacing(1),
	},
}))

const TagGridList: React.FC = () => {
	const classes = useStyles()
	const { dispatch } = useAppContext()

	/** load cache */
	const { data: tags, loading: tagLoading, error: tagError } = useTags()

	if (tagLoading) return <div>loading...</div>
	if (tagError) return <div>failed to load tags.</div>

	/** global state selected tag */
	const handleClick = (event: React.MouseEvent<HTMLDivElement>) => {
		const currentTag = event.currentTarget.getAttribute('aria-label')
		dispatch({
			searchTag: { payload: currentTag },
			prevAction: { payload: 'searchTag' },
		})
	}

	/** scroll control */
	const handleScroll = (positive: boolean) => {
		const scrollAmount = 800
		let toX: number
		if (positive) toX = scrollAmount
		else toX = -1 * scrollAmount

		scroll.scrollMore(toX, {
			horizontal: true,
			containerId: 'chips-wraper',
			smooth: true,
			duration: 400,
		})
	}

	return (
		<>
			<Box className={classes.wraper}>
				<IconButton
					size="small"
					onClick={() => handleScroll(false)}
					data-testid="arrow-icon">
					<ArrowBackIosIcon fontSize="small" />
				</IconButton>
				<Box className={classes.root} id="chips-wraper">
					<Chip
						label="すべて"
						data-testid="chip"
						aria-label="ALL"
						color="primary"
						className={classes.tag}
						onClick={handleClick}
					/>
					{tags.map((tag) => {
						if (!tag.invalid) {
							return (
								<Chip
									key={tag.PK}
									label={tag.name}
									data-testid="chip"
									aria-label={tag.PK}
									color="primary"
									className={classes.tag}
									onClick={handleClick}
								/>
							)
						}
					})}
				</Box>
				<IconButton
					size="small"
					onClick={() => handleScroll(true)}
					data-testid="arrow-icon">
					<ArrowForwardIosIcon fontSize="small" />
				</IconButton>
			</Box>
			<Divider className={classes.divider} />
		</>
	)
}
export default TagGridList
