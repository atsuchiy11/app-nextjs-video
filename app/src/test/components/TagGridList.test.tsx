import * as React from 'react'
import { cleanup, render, fireEvent } from '../testUtils'
import TagGridList from 'src/components/view/TagGridList'

describe('Unit: TagGridList', () => {
	afterEach(cleanup)

	it('matches snapshot', () => {
		const { asFragment } = render(<TagGridList />, {})
		expect(asFragment()).toMatchSnapshot()
	})

	it('show arrow icon for scroll', () => {
		const { getAllByTestId } = render(<TagGridList />, {})
		expect(getAllByTestId('arrow-icon')).toHaveLength(2)
	})

	it('clicked any tag, show alert', () => {
		const { getAllByTestId } = render(<TagGridList />, {})
		// 0~タグ数の範囲で乱数を生成
		const num = Math.floor(
			Math.random() * (getAllByTestId('chip').length + 1)
		)
		const label = getAllByTestId('chip')[num].textContent
		const alert = `タグ「${label}」で絞り込みます`

		window.alert = jest
			.fn()
			.mockImplementationOnce(() => window.alert(alert))

		fireEvent.click(getAllByTestId('chip')[num])
		expect(window.alert).toHaveBeenCalledWith(alert)
	})
})
