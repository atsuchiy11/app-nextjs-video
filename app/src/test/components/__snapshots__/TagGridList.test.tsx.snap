// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit: TagGridList matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root MuiBox-root-5 makeStyles-wraper-1"
  >
    <button
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall"
      data-testid="arrow-icon"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiIconButton-label"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M11.67 3.87L9.9 2.1 0 12l9.9 9.9 1.77-1.77L3.54 12z"
          />
        </svg>
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
    <div
      class="MuiBox-root MuiBox-root-6 makeStyles-root-2"
      id="chips-wraper"
    >
      <div
        class="MuiButtonBase-root MuiChip-root makeStyles-tag-3 MuiChip-clickable"
        data-testid="chip"
        role="button"
        tabindex="0"
      >
        <span
          class="MuiChip-label"
        >
          マーケティング
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </div>
      <div
        class="MuiButtonBase-root MuiChip-root makeStyles-tag-3 MuiChip-clickable"
        data-testid="chip"
        role="button"
        tabindex="0"
      >
        <span
          class="MuiChip-label"
        >
          動画制作
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </div>
      <div
        class="MuiButtonBase-root MuiChip-root makeStyles-tag-3 MuiChip-clickable"
        data-testid="chip"
        role="button"
        tabindex="0"
      >
        <span
          class="MuiChip-label"
        >
          SEM
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </div>
      <div
        class="MuiButtonBase-root MuiChip-root makeStyles-tag-3 MuiChip-clickable"
        data-testid="chip"
        role="button"
        tabindex="0"
      >
        <span
          class="MuiChip-label"
        >
          SEO
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </div>
      <div
        class="MuiButtonBase-root MuiChip-root makeStyles-tag-3 MuiChip-clickable"
        data-testid="chip"
        role="button"
        tabindex="0"
      >
        <span
          class="MuiChip-label"
        >
          HTML
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </div>
      <div
        class="MuiButtonBase-root MuiChip-root makeStyles-tag-3 MuiChip-clickable"
        data-testid="chip"
        role="button"
        tabindex="0"
      >
        <span
          class="MuiChip-label"
        >
          WEBデザイン
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </div>
      <div
        class="MuiButtonBase-root MuiChip-root makeStyles-tag-3 MuiChip-clickable"
        data-testid="chip"
        role="button"
        tabindex="0"
      >
        <span
          class="MuiChip-label"
        >
          SQL
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </div>
      <div
        class="MuiButtonBase-root MuiChip-root makeStyles-tag-3 MuiChip-clickable"
        data-testid="chip"
        role="button"
        tabindex="0"
      >
        <span
          class="MuiChip-label"
        >
          CSS
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </div>
      <div
        class="MuiButtonBase-root MuiChip-root makeStyles-tag-3 MuiChip-clickable"
        data-testid="chip"
        role="button"
        tabindex="0"
      >
        <span
          class="MuiChip-label"
        >
          TypeScript
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </div>
      <div
        class="MuiButtonBase-root MuiChip-root makeStyles-tag-3 MuiChip-clickable"
        data-testid="chip"
        role="button"
        tabindex="0"
      >
        <span
          class="MuiChip-label"
        >
          JavaScript
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </div>
      <div
        class="MuiButtonBase-root MuiChip-root makeStyles-tag-3 MuiChip-clickable"
        data-testid="chip"
        role="button"
        tabindex="0"
      >
        <span
          class="MuiChip-label"
        >
          デジタルツール
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </div>
      <div
        class="MuiButtonBase-root MuiChip-root makeStyles-tag-3 MuiChip-clickable"
        data-testid="chip"
        role="button"
        tabindex="0"
      >
        <span
          class="MuiChip-label"
        >
          WordPress
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </div>
      <div
        class="MuiButtonBase-root MuiChip-root makeStyles-tag-3 MuiChip-clickable"
        data-testid="chip"
        role="button"
        tabindex="0"
      >
        <span
          class="MuiChip-label"
        >
          コンサルティング
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </div>
      <div
        class="MuiButtonBase-root MuiChip-root makeStyles-tag-3 MuiChip-clickable"
        data-testid="chip"
        role="button"
        tabindex="0"
      >
        <span
          class="MuiChip-label"
        >
          若手
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </div>
      <div
        class="MuiButtonBase-root MuiChip-root makeStyles-tag-3 MuiChip-clickable"
        data-testid="chip"
        role="button"
        tabindex="0"
      >
        <span
          class="MuiChip-label"
        >
          リーダー
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </div>
      <div
        class="MuiButtonBase-root MuiChip-root makeStyles-tag-3 MuiChip-clickable"
        data-testid="chip"
        role="button"
        tabindex="0"
      >
        <span
          class="MuiChip-label"
        >
          勤怠
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </div>
      <div
        class="MuiButtonBase-root MuiChip-root makeStyles-tag-3 MuiChip-clickable"
        data-testid="chip"
        role="button"
        tabindex="0"
      >
        <span
          class="MuiChip-label"
        >
          コアバリュー
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </div>
      <div
        class="MuiButtonBase-root MuiChip-root makeStyles-tag-3 MuiChip-clickable"
        data-testid="chip"
        role="button"
        tabindex="0"
      >
        <span
          class="MuiChip-label"
        >
          Yahoo!JAPAN
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </div>
      <div
        class="MuiButtonBase-root MuiChip-root makeStyles-tag-3 MuiChip-clickable"
        data-testid="chip"
        role="button"
        tabindex="0"
      >
        <span
          class="MuiChip-label"
        >
          Google
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </div>
      <div
        class="MuiButtonBase-root MuiChip-root makeStyles-tag-3 MuiChip-clickable"
        data-testid="chip"
        role="button"
        tabindex="0"
      >
        <span
          class="MuiChip-label"
        >
          Criteo
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </div>
      <div
        class="MuiButtonBase-root MuiChip-root makeStyles-tag-3 MuiChip-clickable"
        data-testid="chip"
        role="button"
        tabindex="0"
      >
        <span
          class="MuiChip-label"
        >
          Marketo
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </div>
    </div>
    <button
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall"
      data-testid="arrow-icon"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiIconButton-label"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M5.88 4.12L13.76 12l-7.88 7.88L8 22l10-10L8 2z"
          />
        </svg>
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
  <hr
    class="MuiDivider-root makeStyles-divider-4"
  />
</DocumentFragment>
`;
